![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Guruhlash. `.groupby()`
`.groupby()` meotdi yordamida jadvaldagi ma'lumotlarni biror ustun qiymatlari guruhlarga ajratib tahlil qilishimiz mumkin.

In [17]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/automobile_data.csv", index_col=0)
df.head()

company   body-style  wheel-base  length engine-type  \
index                                                             
0      alfa-romero  convertible        88.6   168.8        dohc   
1      alfa-romero  convertible        88.6   168.8        dohc   
2      alfa-romero    hatchback        94.5   171.2        ohcv   
3             audi        sedan        99.8   176.6         ohc   
4             audi        sedan        99.4   176.6         ohc   

      num-of-cylinders  horsepower  average-mileage    price  
index                                                         
0                 four         111               21  13495.0  
1                 four         111               21  16500.0  
2                  six         154               19  16500.0  
3                 four         102               24  13950.0  
4                 five         115               18  17450.0

Yuqoridagi jadvalda turli avtomobil brandlari haqida ma'lumotlar jamlangan. Keling boshlanishiga har bir brand uchun ustunlarning o'rtacha qiymatini topamiz:

In [6]:
df.groupby('company').max()

body-style  wheel-base  length engine-type num-of-cylinders  \
company                                                                     
alfa-romero    hatchback        94.5   171.2        ohcv              six   
audi               wagon       105.8   192.7         ohc             four   
bmw                sedan       110.0   197.0         ohc              six   
chevrolet          sedan        94.5   158.8         ohc            three   
dodge          hatchback        93.7   157.3         ohc             four   
honda              wagon        96.5   175.4         ohc             four   
isuzu              sedan        94.5   170.7         ohc             four   
jaguar             sedan       113.0   199.6        ohcv           twelve   
mazda              sedan       104.9   175.0       rotor              two   
mercedes-benz      wagon       120.9   208.1        ohcv             five   
mitsubishi         sedan        96.3   172.4         ohc             four   
nissan             wagon       100.4   184.6        ohcv              six   
porsche        hatchback        98.4   175.7        ohcf              six   
toyota             wagon       104.5   187.8         ohc              six   
volkswagen         sedan        97.3   171.7         ohc             four   
volvo              wagon       104.3   188.8         ohc             four   

               horsepower  average-mileage    price  
company                                              
alfa-romero           154               21  16500.0  
audi                  115               24  18920.0  
bmw                   182               23  41315.0  
chevrolet              70               47   6575.0  
dodge                  68               31   6377.0  
honda                 101               30  12945.0  
isuzu                  78               38   6785.0  
jaguar                262               15  36000.0  
mazda                 101               31  18344.0  
mercedes-benz         184               22  45400.0  
mitsubishi             88               37   8189.0  
nissan                152               45  13499.0  
porsche               288               17  37028.0  
toyota                156               35  15750.0  
volkswagen            100               37   9995.0  
volvo                 114               23  13415.0

Yuqoridagi natijani tahlil qilsak, `alfa-romero` mashinalarning o'rtacha narxi (`price`) `15498$`, `honda` mashinalarining o'rtacha ot kuchi (`horsepower`) `92.3` ga teng ekan.

Bitta ustun uchun qiymatlarni olish ham mumkin:

In [7]:
df.groupby('company').price.max() #max narxlarni chiqaramiz

company
alfa-romero      16500.0
audi             18920.0
bmw              41315.0
chevrolet         6575.0
dodge             6377.0
honda            12945.0
isuzu             6785.0
jaguar           36000.0
mazda            18344.0
mercedes-benz    45400.0
mitsubishi        8189.0
nissan           13499.0
porsche          37028.0
toyota           15750.0
volkswagen        9995.0
volvo            13415.0
Name: price, dtype: float64

Guruhlashda bir emas, bir nechta ustunlarni ham ko'rsatish mumkin:

In [20]:
df.groupby(['company','body-style']).min()

wheel-base  length engine-type num-of-cylinders  \
company       body-style                                                     
alfa-romero   convertible        88.6   168.8        dohc             four   
              hatchback          94.5   171.2        ohcv              six   
audi          sedan              99.4   176.6         ohc             five   
              wagon             105.8   192.7         ohc             five   
bmw           sedan             101.2   176.8         ohc             four   
chevrolet     hatchback          88.4   141.1           l             four   
              sedan              94.5   158.8         ohc             four   
dodge         hatchback          93.7   157.3         ohc             four   
honda         sedan              96.5   169.1         ohc             four   
              wagon              96.5   157.1         ohc             four   
isuzu         sedan              94.3   155.9         ohc             four   
jaguar        sedan             102.0   191.7        dohc              six   
mazda         hatchback          93.1   159.1         ohc             four   
              sedan             104.9   175.0         ohc             four   
mercedes-benz hardtop           112.0   199.2        ohcv            eight   
              sedan             110.0   190.9         ohc            eight   
              wagon             110.0   190.9         ohc             five   
mitsubishi    hatchback          93.7   157.3         ohc             four   
              sedan              96.3   172.4         ohc             four   
nissan        sedan              94.5   165.3         ohc             four   
              wagon              94.5   170.2         ohc             four   
porsche       convertible        89.5   168.9        ohcf              six   
              hardtop            89.5   168.9        ohcf              six   
              hatchback          98.4   175.7       dohcv            eight   
toyota        hatchback          95.7   158.7         ohc             four   
              wagon              95.7   169.7        dohc             four   
volkswagen    sedan              97.3   171.7         ohc             four   
volvo         sedan             104.3   188.8         ohc             four   
              wagon             104.3   188.8         ohc             four   

                           horsepower  average-mileage    price  
company       body-style                                         
alfa-romero   convertible         111               21  13495.0  
              hatchback           154               19  16500.0  
audi          sedan               102               18  13950.0  
              wagon               110               19  18920.0  
bmw           sedan               101               15  16430.0  
chevrolet     hatchback            48               38   5151.0  
              sedan                70               38   6575.0  
dodge         hatchback            68               31   6229.0  
honda         sedan               100               24  10345.0  
              wagon                76               30   7295.0  
isuzu         sedan                70               24   6785.0  
jaguar        sedan               176               13  32250.0  
mazda         hatchback            68               17   5195.0  
              sedan                72               31  18344.0  
mercedes-benz hardtop             184               14  45400.0  
              sedan               123               14  25552.0  
              wagon               123               22  28248.0  
mitsubishi    hatchback            68               31   5389.0  
              sedan                88               25   6989.0  
nissan        sedan                55               19   6649.0  
              wagon                69               31   7349.0  
porsche       convertible         207               17  37028.0  
              hardtop   

Jadvaldan xulosa qilsak, `volvo` firmasinig `sedan` turidagi mashinalarining o'rtacha qiymati 12940 `$`    
`wagon`
turidagi mashinalarning o'rtacha narxi esa 13415$.

`.groupby()` yordamida nafaqat o'rta balki boshqa qiymatlarni ham hisoblash mumkin:

In [21]:
df.groupby(['company']).max()

body-style  wheel-base  length engine-type num-of-cylinders  \
company                                                                     
alfa-romero    hatchback        94.5   171.2        ohcv              six   
audi               wagon       105.8   192.7         ohc             four   
bmw                sedan       110.0   197.0         ohc              six   
chevrolet          sedan        94.5   158.8         ohc            three   
dodge          hatchback        93.7   157.3         ohc             four   
honda              wagon        96.5   175.4         ohc             four   
isuzu              sedan        94.5   170.7         ohc             four   
jaguar             sedan       113.0   199.6        ohcv           twelve   
mazda              sedan       104.9   175.0       rotor              two   
mercedes-benz      wagon       120.9   208.1        ohcv             five   
mitsubishi         sedan        96.3   172.4         ohc             four   
nissan             wagon       100.4   184.6        ohcv              six   
porsche        hatchback        98.4   175.7        ohcf              six   
toyota             wagon       104.5   187.8         ohc              six   
volkswagen         sedan        97.3   171.7         ohc             four   
volvo              wagon       104.3   188.8         ohc             four   

               horsepower  average-mileage    price  
company                                              
alfa-romero           154               21  16500.0  
audi                  115               24  18920.0  
bmw                   182               23  41315.0  
chevrolet              70               47   6575.0  
dodge                  68               31   6377.0  
honda                 101               30  12945.0  
isuzu                  78               38   6785.0  
jaguar                262               15  36000.0  
mazda                 101               31  18344.0  
mercedes-benz         184               22  45400.0  
mitsubishi             88               37   8189.0  
nissan                152               45  13499.0  
porsche               288               17  37028.0  
toyota                156               35  15750.0  
volkswagen            100               37   9995.0  
volvo                 114               23  13415.0

`.agg()` metodi yordamida esa birdaniga bir nechta statistik ma'lumotlari olishimiz ham mumkin:

In [22]:
df.groupby('company').agg([len, min, max]) #len - guruhdagi elementlar soni, min - guruh uchun minimum, max - guruh uchun maximum

<ipython-input-22-aef07b916a33>:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  df.groupby('company').agg([len, min, max]) #len - guruhdagi elementlar soni, min - guruh uchun minimum, max - guruh uchun maximum
<ipython-input-22-aef07b916a33>:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df.groupby('company').agg([len, min, max]) #len - guruhdagi elementlar soni, min - guruh uchun minimum, max - guruh uchun maximum


body-style                         wheel-base                \
                     len          min        max        len    min    max   
company                                                                     
alfa-romero            3  convertible  hatchback          3   88.6   94.5   
audi                   4        sedan      wagon          4   99.4  105.8   
bmw                    6        sedan      sedan          6  101.2  110.0   
chevrolet              3    hatchback      sedan          3   88.4   94.5   
dodge                  2    hatchback  hatchback          2   93.7   93.7   
honda                  3        sedan      wagon          3   96.5   96.5   
isuzu                  3        sedan      sedan          3   94.3   94.5   
jaguar                 3        sedan      sedan          3  102.0  113.0   
mazda                  5    hatchback      sedan          5   93.1  104.9   
mercedes-benz          4      hardtop      wagon          4  110.0  120.9   
mitsubishi             4    hatchback      sedan          4   93.7   96.3   
nissan                 5        sedan      wagon          5   94.5  100.4   
porsche                3  convertible  hatchback          3   89.5   98.4   
toyota                 7    hatchback      wagon          7   95.7  104.5   
volkswagen             4        sedan      sedan          4   97.3   97.3   
volvo                  2        sedan      wagon          2  104.3  104.3   

              length               engine-type  ... num-of-cylinders  \
                 len    min    max         len  ...              max   
company                                         ...                    
alfa-romero        3  168.8  171.2           3  ...              six   
audi               4  176.6  192.7           4  ...             four   
bmw                6  176.8  197.0           6  ...              six   
chevrolet          3  141.1  158.8           3  ...            three   
dodge              2  157.3  157.3           2  ...             four   
honda              3  157.1  175.4           3  ...             four   
isuzu              3  155.9  170.7           3  ...             four   
jaguar             3  191.7  199.6           3  ...           twelve   
mazda              5  159.1  175.0           5  ...              two   
mercedes-benz      4  190.9  208.1           4  ...             five   
mitsubishi         4  157.3  172.4           4  ...             four   
nissan             5  165.3  184.6           5  ...              six   
porsche            3  168.9  175.7           3  ...              six   
toyota             7  158.7  187.8           7  ...              six   
volkswagen         4  171.7  171.7           4  ...             four   
volvo              2  188.8  188.8           2  ...             four   

              horsepower           average-mileage         price           \
                     len  min  max             len min max   len      min   
company                                                                     
alfa-romero            3  111  154               3  19  21     3  13495.0   
audi                   4  102  115               4  18  24     4  13950.0   
bmw                    6  101  182               6  15  23     6  16430.0   
chevrolet              3   48   70               3  38  47     3   5151.0   
dodge                  2   68   68               2  31  31     2   6229.0   
honda                  3   76  101               3  24  30     3   7295.0   
isuzu                  3   70   78               3  24  38     3   6785.0   
jaguar                 3  176  262               3  13  15     3  32250.0   
mazda                  5   68  101               5  17  31     5   5195.0   
mercedes-benz          4  123  184               4  14  22     4  25552.0   
mitsubishi             4   68   88               4  25  37     4   5389.0   
nissan                 5   55  152               5  19  45     5   6649.0   
porsche                3  

`.agg()` yordamida istalgan funksiyalarni chaqirishimiz mumkin, misol uchun `numpy` dagi `mean` (o'rta qiymat) funiksiyasini ham qo'shamiz. Natijalarni faqat `price` ustuni uchun chiqaramiz:

In [23]:
df.groupby('body-style').price.agg([len, min, max, np.mean])

<ipython-input-23-054e310ab442>:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  df.groupby('body-style').price.agg([len, min, max, np.mean])
<ipython-input-23-054e310ab442>:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df.groupby('body-style').price.agg([len, min, max, np.mean])
<ipython-input-23-054e310ab442>:1: FutureWarning: The provided callable <function mean at 0x7e3e8add9da0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df.groupby('body-style').price.agg([len, min, max, np.mean])


len      min      max          mean
body-style                                      
convertible    3  13495.0  37028.0  22341.000000
hardtop        2  34028.0  45400.0  39714.000000
hatchback     15   5151.0  16500.0   7159.571429
sedan         32   6575.0  41315.0  17706.333333
wagon          9   6918.0  28248.0  12730.111111

`.apply()` metodi yordamida guruhdagi qiymatlarni o'zgartirishimiz ham mumkin: